File: backend/src/Analytics.cpp
#include "Analytics.h"

#include <algorithm>

namespace Analytics {

std::optional<double> movingAverage(const std::vector<Tick>& window) {
  if (window.empty()) return std::nullopt;
  double sum = 0.0;
  for (const auto& t : window) sum += t.price;
  return sum / static_cast<double>(window.size());
}

std::optional<double> vwap(const std::vector<Tick>& window) {
  if (window.empty()) return std::nullopt;
  long double pv = 0.0L;
  long double v = 0.0L;
  for (const auto& t : window) {
    pv += static_cast<long double>(t.price) * static_cast<long double>(t.volume);
    v += static_cast<long double>(t.volume);
  }
  if (v == 0.0L) return std::nullopt;
  return static_cast<double>(pv / v);
}

std::optional<Ohlc> ohlc(const std::vector<Tick>& window) {
  if (window.empty()) return std::nullopt;
  Ohlc o{};
  o.open = window.front().price;
  o.close = window.back().price;
  auto [minIt, maxIt] = std::minmax_element(
      window.begin(), window.end(), [](const Tick& a, const Tick& b) { return a.price < b.price; });
  o.low = minIt->price;
  o.high = maxIt->price;
  return o;
}

} // namespace Analytics

File: api/src/analytics.js
// Analytics helpers for MA, VWAP, and OHLC over the last N ticks.
// These computations are windowed by "last N ticks" for a straightforward demo.

/**
 * @param {Array<{price:number}>} window
 */
function movingAverage(window) {
  if (!window.length) return null;
  const sum = window.reduce((acc, t) => acc + t.price, 0);
  return sum / window.length;
}

/**
 * @param {Array<{price:number, volume:number}>} window
 */
function vwap(window) {
  if (!window.length) return null;
  let pv = 0;
  let v = 0;
  for (const t of window) {
    pv += t.price * t.volume;
    v += t.volume;
  }
  if (v === 0) return null;
  return pv / v;
}

/**
 * @param {Array<{price:number}>} window
 */
function ohlc(window) {
  if (!window.length) return null;
  const open = window[0].price;
  const close = window[window.length - 1].price;
  let high = -Infinity;
  let low = Infinity;
  for (const t of window) {
    high = Math.max(high, t.price);
    low = Math.min(low, t.price);
  }
  return { open, high, low, close };
}

module.exports = { movingAverage, vwap, ohlc };

